### DISCUSSION TITLE
Generate vim syntax file

### DISCUSSION THREAD

**Discussion by alltilla** [alltilla/axosyslog-cfg-helper#1] [Ideas]
Based on @smortex's idea: https://github.com/syslog-ng/syslog-ng/pull/4475#issuecomment-1546775895

> Nice! I am wondering if it would make sense to likewise remove the vim syntax file I recently updated: they are only part of the RPM packages, and for vim users, it is more convenient to have a repository to clone using pathogen / plugged.
>
> We could setup a dedicated repo, e.g. syslog-ng/vim-syslog-ng where the main branch has the actual vim config and another branch has the tooling to regenerate new config using your tool when the syntax of syslog-ng is extended.

**Comment by alltilla** [comment]
I think this is definitely a good idea.

IMO it would be best to generate the syntax file from released versions of [syslog-ng-cfg-helper](https://github.com/alltilla/syslog-ng-cfg-helper).

#### Pros
  * We do not have to checkout the syslog-ng source, and build a tarball.
  * We do not have to setup an env for generating the option database, which is a bit tricky because of `bison`.
  * We do not have to generate the option database, so we can save writing a couple of scripts to do so.

#### Cons
  * We depend on the release of syslog-ng-cfg-helper.
    * This is manual right now, but I can make it work automatically for each new syslog-ng release.
    * It only gets released for stable syslog-ng releases, so the syntax file does not get updated for the changes in-between.

**Comment by alltilla** [comment]
Another thing is that I don't think that we can/should develop the vim syntax generating project under the syslog-ng namespace/organization. I don't know if the the admin of the https://github.com/syslog-ng organization would let us create a new project there.

**Comment by smortex** [comment]
Awesome @alltilla :heart_eyes: 

I started some proof-of-concept repo under my profile and am willing to transfer it to @syslog-ng if it succeeds to reach its goal: https://github.com/smortex/vim-syslog-ng.  For the sake of simplicity, for now I just used a template and fill-in this template with the information from the .db file generated by `syslog-ng-cfg-helper`.  I used Ruby because it is what I am most proficient with, but have nothing against a rewrite in Python if it helps in some way.

I have noted a few inconsistencies with the current syntax where some keywords are not in the db file:
* `options { threaded(yes); }` => threaded
* `filter f_authpriv { facility(authpriv); };` => facility
* `rewrite { set-tag('syslog'); };` => set-tag (and all tag-related functions)

`KW_LOWER` and `KW_UPPER` are also listed as keyword, but I guess this was a side-effect fixed by https://github.com/syslog-ng/syslog-ng/commit/9399cad3e7081c98d1b5a08cb2d075276ef340e1.

I kept these inconsistencies "as is" for now.

**Comment by alltilla** [comment]
This looks really cool :)

I can hopefully continue this week with:
  * Fixing the missing/incomplete keywords you have mentioned.
  * Writing some GitHub Actions workflows which tracks the releases of syslog-ng-cfg-helper, generates the syntax file and pushes it to the main branch.

Although I am not really familiar with Ruby, we can keep using it. The only thing that could be useful when using python is that I have [classes implemented](https://github.com/alltilla/syslog-ng-cfg-helper/tree/master/syslog_ng_cfg_helper/driver_db) for the database, and we could use/extend those if we need more sophisticated traversal. But I can imagine that a simple one that you have written will take us to great lengths.

**Comment by smortex** [comment]
Nice! I sent you an invitation on the vim-syslog-ng project so that we can work together more easily. Feel free to open PR and ask me to add commits to update the Ruby code if that helps you!  And indeed if it becomes tedious we will rewrite the generator!

**Comment by alltilla** [comment]
The project is available at https://github.com/syslog-ng/vim-syslog-ng
